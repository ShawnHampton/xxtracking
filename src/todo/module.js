import { fromJS } from 'immutable';
import { createSelector } from 'reselect';

export const CONSTANTS = {
	addTodo: 'todo/addTodo',
	addTodoAsync: 'todo/addTodoAsync',
	removeTodo: 'todo/removeTodo'
};

///-- ACTIONS --///

export const addTodo = ({ id, text }) => ({
	type: CONSTANTS.addTodo,
	payload: { id, text }
});
export const removeTodo = id => ({
	type: CONSTANTS.removeTodo,
	payload: { id }
});

const initialState = {};

const reducer = (state = fromJS(initialState), action) => {
	const { type, payload } = action;
	switch (type) {
		case CONSTANTS.addTodo:
		case CONSTANTS.addTodoAsync:
			return state.merge({
				[payload.id]: {
					text: payload.text
				}
			});
		case CONSTANTS.removeTodo:
			return state.delete(payload.id);
		default:
			return state;
	}
};

export default reducer;

///-- SELECTORS --///
const getTodosImmutable = state => state.get('todos', {});
const getTodoImmutable = (state, props) => state.getIn(['todos', props.id], {});

export const getTodoIds = createSelector([getTodosImmutable], todos =>
	todos.keySeq().toArray()
);
export const getTodo = createSelector([getTodoImmutable], todo =>
	todo.toObject()
);
