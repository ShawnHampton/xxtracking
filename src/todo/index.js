import List, { List as ListRaw } from './list';
import Todo, { Todo as TodoRaw } from './todo';
import module from './module';

export { List, ListRaw, Todo, TodoRaw, module };

export default List;
