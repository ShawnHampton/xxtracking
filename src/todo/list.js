import React from 'react';
import Todo from 'todo/todo';
import { connect } from 'react-redux';
import { getTodoIds, addTodo } from 'todo/module';
import { uniqueId } from 'lodash/fp';

export class List extends React.PureComponent {
	state = {
		text: ''
	};
	onAddTodo = () => {
		this.props.addTodo({ id: uniqueId(), text: this.state.text });
	};
	onTextChange = event => this.setState({ text: event.target.value });
	render() {
		const { todoIds } = this.props;
		return (
			<main>
				<h1>Todo List</h1>
				<section>{todoIds.map(id => <Todo key={id} id={id} />)}</section>
				<section>
					<input type="text" onChange={this.onTextChange} />
					<button className="button" onClick={this.onAddTodo}>
						Add Todo
					</button>
				</section>
			</main>
		);
	}
}

const mapStateToProps = state => ({
	todoIds: getTodoIds(state)
});

export default connect(
	mapStateToProps,
	{
		addTodo
	}
)(List);
