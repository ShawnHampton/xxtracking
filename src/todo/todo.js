import React from 'react';
import { connect } from 'react-redux';
import { getTodo, removeTodo } from 'todo/module';

export class Todo extends React.PureComponent {
	onRemove = () => this.props.removeTodo(this.props.id);
	render() {
		const { text } = this.props;
		return (
			<article>
				<span>{text}</span>
				<button onClick={this.onRemove}>Remove</button>
			</article>
		);
	}
}

const mapStateToProps = (state, props) => ({
	text: getTodo(state, props).text
});

export default connect(
	mapStateToProps,
	{
		removeTodo
	}
)(Todo);
