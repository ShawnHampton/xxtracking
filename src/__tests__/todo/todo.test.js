import React from 'react';
import { shallow } from 'enzyme';
// Grabbing only the unconnected raw component
import { TodoRaw as Todo } from 'todo';

const config = {
	text: 'test'
};

function makeStr(length = 10) {
	let text = '';
	const possible =
		'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	for (let i = 0; i < length; i++)
		text += possible.charAt(Math.floor(Math.random() * possible.length));
	return text;
}
const randStr = makeStr(10000);
const onTextChangeCases = [
	{
		name: 'Basic Change',
		value: '+A+,,SDF12.34.56',
		result: '+A+,,SDF12.34.56'
	},
	{
		name: 'Weird Characters',
		value: 'Af02h$#480kj~```(cÕ„®)',
		result: 'Af02h$#480kj~```(cÕ„®)'
	},
	{
		name: 'No Characters',
		value: '',
		result: ''
	},
	{
		name: 'Random very long Alphanumeric string',
		value: randStr,
		result: randStr
	}
];

describe('Todo', () => {
	let wrapper = null;
	beforeEach(() => {
		wrapper = shallow(<Todo {...config} />);
	});

	describe('Setup and Props', () => {
		test('Match previous snapshot', () => {
			expect(wrapper).toMatchSnapshot();
		});
		test('Match config props', () => {
			expect(wrapper.instance().props).toEqual(expect.objectContaining(config));
		});
	});

	describe('User Input Cases', () => {
		describe('On Text Change', () => {
			onTextChangeCases.forEach(item => {
				test(item.name, () => {
					const value = item.value;
					wrapper.setProps({ ...item.props, text: value });
					let text = wrapper
						.find('article')
						.find('span')
						.text();
					expect(text).toEqual(item.result);
				});
			});
		});
	});
});
