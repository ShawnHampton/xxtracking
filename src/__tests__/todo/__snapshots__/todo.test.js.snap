// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Setup and Props Match previous snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Todo
    text="test"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span>
          test
        </span>,
        <button
          onClick={[Function]}
        >
          Remove
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "test",
        },
        "ref": null,
        "rendered": "test",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Remove",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Remove",
        "type": "button",
      },
    ],
    "type": "article",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span>
            test
          </span>,
          <button
            onClick={[Function]}
          >
            Remove
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "test",
          },
          "ref": null,
          "rendered": "test",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Remove",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Remove",
          "type": "button",
        },
      ],
      "type": "article",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
